app/models/client.rb

```diff
@@ -7,14 +7,32 @@ def full_name
   end
 
   def contacted_at
-    # the date of the most recent message sent or received
+    # the date of the most recent message sent to or received from this client
     last_message = Message.where(client: self).order(:created_at).last
     if last_message
       return last_message.created_at
     end
     self.updated_at
   end
 
+  def received_message_at
+    # the date of the most recent message received from this client
+    last_received_message = Message
+      .where(client: self, inbound: true)
+      .order(:created_at)
+      .last
+
+    if last_received_message
+      return last_received_message.created_at
+    end
+    self.updated_at
+  end
+
+  def unread_message_count
+    # the number of messages received that are unread
+    Message.where(read: false, client: self).count
```

Comment:
A more reusable approach would be to define a scope for this:

```ruby
# in Messages model:
scope :unread, -> { where(read: false) }

# In Client model
def unread_message_count
  messages.unread.count
end
```

In general, using associations and scopes is more flexible because it allows you to assign semantic meaning to certain queries - so if the implementation of unread messages changes later, it's easy to tell which queries need to change.

It also allows you to compose queries alongside the query DSL, for example:

```
client.messages.unread.sent_by(Devices::PHONE).order(created_at: :desc)
```

Also using associations and scopes allows for more performance optimizations by using activerecord preloader and includes. In general, scopes are awesome and almost all queries should be using a scope or association.